src/gluonts/model/n_beats/_network.py

```diff
@@ -0,0 +1,320 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import List
+
+# Third-party imports
+import mxnet as mx
+import numpy as np
+
+# First-party imports
+from gluonts.block.scaler import MeanScaler, NOPScaler
+from gluonts.core.component import validated
+from gluonts.distribution import Distribution, DistributionOutput
+from gluonts.model.common import Tensor
+from typing import List, Optional, Tuple
+
+VALID_N_BEATS_STACK_TYPES = "G", "S", "T"
+
+
+def linear_space(F, backcast_length, forecast_length, fwd_looking=True):
+    ls = (
+        F.arange(-float(backcast_length), float(forecast_length), 1)
+        / backcast_length
+    )
+    if fwd_looking:
+        ls = ls[backcast_length:]
+    else:
+        ls = ls[:backcast_length]
+    return ls
+
+
+def seasonality_model(
+    F,
+    thetas,
+    expansion_coefficient_length,
+    context_length,
+    prediction_length,
+    is_forecast,
+):
+    p = expansion_coefficient_length
+    p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
+    t = linear_space(
+        F, context_length, prediction_length, fwd_looking=is_forecast
+    )
+    s1 = F.stack(*[F.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)
+    s2 = F.stack(*[F.sin(2 * np.pi * i * t) for i in range(p2)], axis=0)
+    S = F.concat(s1, s2, dim=0)
+    S = F.cast(S, np.float32)
+    return F.dot(thetas, S)
+
+
+def trend_model(
+    F,
+    thetas,
+    expansion_coefficient_length,
+    context_length,
+    prediction_length,
+    is_forecast,
+):
+    p = expansion_coefficient_length
+    t = linear_space(
+        F, context_length, prediction_length, fwd_looking=is_forecast
+    )
+    T = F.transpose(F.stack(*[t ** i for i in range(p)], axis=0))
+    T = F.cast(T, np.float32)
+    return F.dot(thetas, F.transpose(T))
```

Comment:
I have basically used the exact same approach as here: https://github.com/philipperemy/n-beats/blob/master/nbeats_pytorch/model.py
This implementation contradicts the paper in one point by always dividing by backcast_length (even for forecast).
However, the n-beats paper is unfortunately under-specifying how to implement this part regarding the backcast and this implementation seems reasonable.

Reply:
I tryed couple of different ways of implementing this, they dont seem to drastically vary in performance. Also I reworked the whole seasonal and trend model implementation.